File: app/checks/html.py
"""What the raw HTML says - and nothing about what JavaScript would add.

The one honesty rule in this file: `checked` is always reported, and its value is
"raw_html_only". Plenty of sites inject <title> and <meta name=viewport> from
JavaScript, so "not present in the raw HTML" is a different claim from "missing",
and the API must not blur the two. Every issue this module raises is named
..._IN_RAW_HTML for the same reason.

Mixed content is counted strictly: only sub-resources a browser would actually
block or warn about - img/script/iframe src and link rel=stylesheet href - and
only when the page itself was served over https.
"""
from __future__ import annotations

from urllib.parse import urlsplit

from selectolax.parser import HTMLParser

RAW_ONLY = "raw_html_only"
TRUNCATED = "truncated_2mb"
NOT_HTML = "not_html"
MAX_SAMPLES = 5


def _is_insecure(value: str) -> bool:
    return value.strip().lower().startswith("http://")


def parse(body: bytes, *, final_url: str, content_type: str,
          truncated: bool = False) -> dict:
    """The `html` block of a check result."""
    checked = TRUNCATED if truncated else RAW_ONLY
    base_is_https = urlsplit(final_url).scheme.lower() == "https"

    if content_type and "html" not in content_type.lower():
        return {"checked": NOT_HTML, "content_type": content_type,
                "title": {"present": False}, "viewport": {"present": False},
                "mixed_content": {"count": 0, "samples": []},
                "stylesheets": 0, "scripts": 0}

    tree = HTMLParser(body.decode("utf-8", errors="replace"))

    title_node = tree.css_first("title")
    title_text = (title_node.text(strip=True) if title_node else "") or ""
    title = {"present": bool(title_text), "text": title_text} if title_text else {"present": False}

    viewport = {"present": False}
    for meta in tree.css("meta"):
        if (meta.attributes.get("name") or "").strip().lower() == "viewport":
            content = (meta.attributes.get("content") or "").strip()
            viewport = {"present": True, "content": content} if content else {"present": True}
            break

    insecure: list[str] = []
    for node in tree.css("img, script, iframe"):
        src = node.attributes.get("src") or ""
        if src and _is_insecure(src):
            insecure.append(src)

    stylesheets = 0
    for node in tree.css("link"):
        rel = (node.attributes.get("rel") or "").strip().lower()
        if "stylesheet" not in rel.split():
            continue
        stylesheets += 1
        href = node.attributes.get("href") or ""
        if href and _is_insecure(href):
            insecure.append(href)

    scripts = len(tree.css("script"))

    if not base_is_https:
        # http:// sub-resources on an http:// page are not "mixed" content.
        insecure = []

    return {
        "checked": checked,
        "title": title,
        "viewport": viewport,
        "mixed_content": {"count": len(insecure), "samples": insecure[:MAX_SAMPLES]},
        "stylesheets": stylesheets,
        "scripts": scripts,
    }

File: src/checks/html.ts
/**
 * What the raw HTML says - and nothing about what JavaScript would add.
 *
 * The one honesty rule in this file: `checked` is always reported, and its value
 * is "raw_html_only". Plenty of sites inject <title> and <meta name=viewport>
 * from JavaScript, so "not present in the raw HTML" is a different claim from
 * "missing", and the API must not blur the two. Every issue this module feeds is
 * named ..._IN_RAW_HTML for the same reason.
 *
 * Parsing is HTMLRewriter, not a DOM library. It is native and streaming, which
 * is what keeps a 2 MB page inside the Worker's 10 ms CPU budget - a JS parse of
 * the same page would not fit.
 *
 * Mixed content is counted strictly: only sub-resources a browser would actually
 * block - img/script/iframe src and link rel=stylesheet href - and only when the
 * page itself was served over https.
 */

export const RAW_ONLY = "raw_html_only";
export const TRUNCATED = "truncated_2mb";
export const NOT_HTML = "not_html";
export const MAX_SAMPLES = 5;

export interface HtmlReport {
  checked: string;
  content_type?: string;
  title: { present: boolean; text?: string };
  viewport: { present: boolean; content?: string };
  mixed_content: { count: number; samples: string[] };
  stylesheets: number;
  scripts: number;
}

const isInsecure = (value: string): boolean => value.trim().toLowerCase().startsWith("http://");

export async function parseHtml(
  body: Uint8Array,
  opts: { finalUrl: string; contentType: string; truncated?: boolean },
): Promise<HtmlReport> {
  const checked = opts.truncated ? TRUNCATED : RAW_ONLY;

  if (opts.contentType && !opts.contentType.toLowerCase().includes("html")) {
    return {
      checked: NOT_HTML,
      content_type: opts.contentType,
      title: { present: false },
      viewport: { present: false },
      mixed_content: { count: 0, samples: [] },
      stylesheets: 0,
      scripts: 0,
    };
  }

  let baseIsHttps = false;
  try {
    baseIsHttps = new URL(opts.finalUrl).protocol === "https:";
  } catch {
    baseIsHttps = false;
  }

  let titleText = "";
  let inTitle = false;
  let viewportSeen = false;
  let viewportContent = "";
  let stylesheets = 0;
  let scripts = 0;
  const insecure: string[] = [];

  const note = (value: string | null) => {
    if (value && isInsecure(value)) insecure.push(value);
  };

  const rewriter = new HTMLRewriter()
    .on("title", {
      element() {
        // Only the first <title> counts; a browser ignores later ones.
        inTitle = titleText === "";
      },
      text(chunk) {
        if (inTitle) titleText += chunk.text;
      },
    })
    .on("meta", {
      element(el) {
        if ((el.getAttribute("name") ?? "").trim().toLowerCase() === "viewport" && !viewportSeen) {
          viewportSeen = true;
          viewportContent = (el.getAttribute("content") ?? "").trim();
        }
      },
    })
    .on("img, script, iframe", {
      element(el) {
        note(el.getAttribute("src"));
      },
    })
    .on("script", {
      element() {
        scripts += 1;
      },
    })
    .on("link", {
      element(el) {
        const rel = (el.getAttribute("rel") ?? "").trim().toLowerCase();
        if (!rel.split(/\s+/).includes("stylesheet")) return;
        stylesheets += 1;
        note(el.getAttribute("href"));
      },
    });

  // Transforming drives the parse; the output is thrown away.
  await rewriter.transform(new Response(body)).arrayBuffer();

  const title = titleText.trim();
  // An http:// page loading http:// sub-resources is not "mixed" content.
  const mixed = baseIsHttps ? insecure : [];

  return {
    checked,
    title: title ? { present: true, text: title } : { present: false },
    viewport: viewportSeen
      ? viewportContent
        ? { present: true, content: viewportContent }
        : { present: true }
      : { present: false },
    mixed_content: { count: mixed.length, samples: mixed.slice(0, MAX_SAMPLES) },
    stylesheets,
    scripts,
  };
}
